### ISSUE TITLE
network: start nodes at user-defined position, but don't fix

### ISSUE THREAD

**Issue by klmdb** [almende/vis#711]
When I set the node's x and y attributes before adding it to my network's Dataset, the nodes are initialized at that position, but they can't be moved anymore. I would like to be able to give them a starting position from where they continue to stabilize. 

For now, I modified the Node.js source file to achieve this functionality:

CHANGE

    if (properties.allowedToMoveX !== undefined) {
      this.xFixed = !properties.allowedToMoveX;
      this.allowedToMoveX = properties.allowedToMoveX;
    }
    else if (properties.x !== undefined && this.allowedToMoveX == false) {
      this.xFixed = true;
    }

    if (properties.allowedToMoveY !== undefined) {
      this.yFixed = !properties.allowedToMoveY;
      this.allowedToMoveY = properties.allowedToMoveY;
    }
    else if (properties.y !== undefined && this.allowedToMoveY == false) {
      this.yFixed = true;
    }

FOR

    if (properties.allowedToMoveX !== undefined) {
      this.xFixed = !properties.allowedToMoveX;
      this.allowedToMoveX = properties.allowedToMoveX;
    }
    else if (properties.x !== undefined && this.allowedToMoveX == false) {
      this.xFixed = true;
    }

    if (properties.allowedToMoveY !== undefined) {
      this.yFixed = !properties.allowedToMoveY;
      this.allowedToMoveY = properties.allowedToMoveY;
    }
    else if (properties.y !== undefined && this.allowedToMoveY == false) {
      this.yFixed = true;
    }

    if(properties.startX !== undefined){

      if(this.x === null){
        this.x = properties.startX;
      }
    }
    if(properties.startY !== undefined){

      if(this.y === null){
        this.y = properties.startY;
      }
    }

**Comment by AlexDM0** [comment]
Hi,

Why didn't you just add the allowedToMoveX: true and same for Y to the node options? That should do exactly what you want. 

Regards

**Comment by klmdb** [comment]
Thanks, that does help me. However, setting the x and y attributes on a node causes it to be set to those coordinates every time I call DataSet.update(node) (which I use to update a bunch of attributes I keep in my VisWrapper). I don't see a clear way to only set the positions once, except for updating all the nodes with the x and y attributes and then clearing tose attributes right after. 

Would it be possible/desireable to provide a node.setPosition() method that sets the position only once, when it is called?

**Comment by klmdb** [comment]
On a related node, when I initialize the network with the nodes' positions set (in a non-stabilized configuration), the stabilization is frozen. That is to say, the positions of the nodes doesn't change until I drag one of them and the ontire graph starts stabilizing again.

**Comment by AlexDM0** [comment]
Hi,

I see, @josdejong could you look into this? This might be one reason the dataset was used differently in the network as we discussed previously. 

@klmdb, if you do not change the x and y values with your update to the dataset, it should not snap back to the position you gave it earlier. That's what you get now right?

It SHOULD work like this: give x and y => lock in that position. If you also define the allowedToMoves it should just move. Any update later on should not fix it again unless you specifically disable the allowedToMoves. Nor should any update to an unrelated field snap them back. 

Regards,

Alex

**Comment by AlexDM0** [comment]
Could you make a JSBin to show this?

Regards

**Comment by klmdb** [comment]
Alex, I don't have the time right now to create a JSBin, but I'll try to elaborate a little more.

Regarding the node snapping back to the x,y position on update, that's probably a misuse of the library on my part. I set the x and y attributes on the node on initialization, and it still has those values set when I update the node later on, so the update resets it. I'm assuming this is because the version of node that I use in my scope has it's own value for the x and y attributes, which are not updated by vis, and therefore, after the node has moved because of stabilization, when I update the node, which still has thze original x and y positions, it snaps back.

I could probably quick-solve this by removing the x and y attributes on my version of the node after it has been added to the dataset on initialization. This messes up my code a little bit though, so I'd prefere to be able to just set it once and be done with it (which I solved with startX and startY).

**Comment by klmdb** [comment]
Do you guys have a basic usage JSBin example for vis? I might be able to throw together a quick example if I could start from that.

**Comment by AlexDM0** [comment]
Hi Klmdb,

Since x and y are properties of the node that you can set, we are not going to jump through hoops to give these properties special requirements. As I see it, the API should (if functioning correctly) do what you want. If you tell the dataset to snap your nodes to a position, it will do so and I don't plan on changing that just because it interferes with your implementation :)

As for the jsBin, there's no rush, you can use the code of our examples as a starting point. 

Regards,

Alex

**Comment by AlexDM0** [comment]
Hi,

Is this still an issue? We just released v4 with a lot of changes on this part. If you supply an x and y coordinate the nodes will spawn there (or snap there if already initialized) but they will not be fixed unless you supply the fixed option. I'll close this issue as most of what we discussed here is no longer relevant with v4. If you encounter a similar problem in v4, please make a new issue.

Regards,

Alex
